pdns/pdns_recursor.cc

```diff
@@ -4078,125 +4077,130 @@ static void handleRCC(int fd, FDMultiplexer::funcparam_t& var)
   }
 }
 
-static void handleTCPClientReadable(int fd, FDMultiplexer::funcparam_t& var)
+static void TCPIOHandlerIO(int fd, FDMultiplexer::funcparam_t& var)
 {
-  PacketID* pident=boost::any_cast<PacketID>(&var);
-  //  cerr<<"handleTCPClientReadable called for fd "<<fd<<", pident->inNeeded: "<<pident->inNeeded<<", "<<pident->sock->getHandle()<<endl;
-
-  boost::shared_array<char> buffer(new char[pident->inNeeded]);
+  auto pid = boost::any_cast<std::shared_ptr<PacketID>>(var);
+  assert(pid->tcphandler);
+  assert(fd == pid->tcphandler->getDescriptor());
+  IOState newstate = IOState::Done;
 
-  ssize_t ret=recv(fd, buffer.get(), pident->inNeeded,0);
-  if(ret > 0) {
-    pident->inMSG.insert(pident->inMSG.end(), &buffer[0],  &buffer[ret]);
-    pident->inNeeded -= (size_t)ret;
-    if(!pident->inNeeded || pident->inIncompleteOkay) {
-      //      cerr<<"Got entire load of "<<pident->inMSG.size()<<" bytes"<<endl;
-      PacketID pid=*pident;
-      PacketBuffer msg = pident->inMSG;
+  TCPLOG("TCPIOHandlerIO: lowState " << int(pid->lowState) << endl);
 
-      t_fdm->removeReadFD(fd);
-      MT->sendEvent(pid, &msg);
+  switch (pid->highState) {
+  case TCPAction::DoingRead:
+    TCPLOG("highState: Reading" << endl);
+    // try reading
+    try {
+      size_t pos = pid->inMSG.size();
+      pid->inMSG.resize(pos + pid->inNeeded); // make room for what we'll read
+      newstate = pid->tcphandler->tryRead(pid->inMSG, pos, pid->inNeeded);
+      switch (newstate) {
+      case IOState::Done:
+      case IOState::NeedRead:
+        TCPLOG("tryRead: Done or NeedRead " << int(newstate) << ' ' << pos << '/' << pid->inNeeded << endl);
+        pid->inMSG.resize(pos); // old content (if there) + new bytes read
+        pid->inNeeded -= pos;
+        TCPLOG("TCPIOHandlerIO " << pid->inNeeded << ' ' << pid->inIncompleteOkay << endl);
+        if (pid->inNeeded == 0 || pid->inIncompleteOkay) {
+          newstate = IOState::Done;
+          // removeReadFD seems to clobber PacketID, so take a copy
```

Comment:
That seems weird, shouldn't `PacketID` be opaque to the multiplexer?

Reply:
I suppose this is a remainder of my earlier confusion. Let me clean up this and the other issues, it *should* lead to cleaner code.